Refactoring in preparation of adding support for test names

netdpb opened the pull request:
Extracted `Fact`. `ExpectedFact`, and `ReportedFact` to top level types, and adjusted their accessibility.

Added `ExpectedFact.Reader` to contain logic to read expected facts from files, and `ConformanceTestReport.Builder` to build up a report.

Added unit tests for `ExpectedFact.Reader`.

Preparation for fixing #141.

netdpb committed:
Make `Fact`, `ExpectedFact`, and `ReportedFact` top-level classes, and reduce their accessibility.
Make `ReportedFact.matches()` take `ExpectedFact`.

netdpb committed:
Add `ConformanceTestReport.Builder` and `ExpectedFact.Reader`. Add unit tests for `ExpectedFact.Reader`.

netdpb committed:
Fixups.

wmdietl commented on conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java:
```diff
@@ -0,0 +1,167 @@
+                    + ")")
+                + "\\s*");
+
+    private final Map<Long, String> facts = new HashMap<>();
```
These two are fields to avoid the allocation of a new `HashMap` for multiple calls of `readExpectedFacts` on the same `Reader`?

netdpb replied:
Sort of. they don't need to be fields yet, but it'll help when I add support for test names. I'll remove them from this.

wmdietl commented on conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java:
```diff
@@ -0,0 +1,167 @@
+            facts.put(lineNumber, expectation.trim());
+          }
+        } else {
+          facts.forEach(
```
This assumes that the last line is not matching `EXPECTATION_COMMENT`, right? Otherwise, `facts` might contain things that will never be added to `expectedFacts.
A comment might help here, or an assertion before the `return` that `facts` is empty.

netdpb replied:
I've added a TODO. The upcoming PR to deal with test names will include error handling.

wmdietl commented on conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java:
```diff
@@ -0,0 +1,167 @@
+// Copyright 2023 The JSpecify Authors
```
Should a new file use `2024` or is it more important that some of the logic in this file existed in 2023 already?

netdpb replied:
This is mostly just extracting code from one file into another. so I kept the copyright year. I see guidance that says not to change the copyright year when moving or copying a file. You could consider this copying-then-editing.

But if someone has more specific guidance, I'm happy to follow it.

wmdietl commented on conformance-test-framework/src/main/java/org/jspecify/conformance/ExpectedFact.java:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  /**
+   * Returns an expected fact representing that the source type cannot be converted to the sink type
```
From this description I expected the return type to be `ExpectedFact`. Is there a better term for the returned String?

netdpb replied:
Yes. Based on this I've tried to use "factText" throughout, and merged `ReportedFact.expectedFact()` into `ReportedFact.getFactText()`.

I also noticed that these static methods are only used from implementations of `ReportedFact`, so I moved them there.

wmdietl reviewed (commented):
I have a few questions.

netdpb committed:
Consistently use "factText" to refer to the assertion in a comment.
Merge `ReportedFact.expectedFact()` with `ReportedFact.getFactText()`.
Move the static methods that create fact text into `ReportedFact` since they're meant to be called by its implementations.

netdpb committed:
Add TODO to check if there are any unemitted facts at the end of the file.

netdpb reviewed (commented):
Thanks.

netdpb committed:
Small refactoring.

netdpb committed:
Small cleanups.

wmdietl reviewed (approved):
Thanks, looks good!
